ddtrace/internal/module.py

```diff
@@ -0,0 +1,430 @@
+        # Module is porbably only partially initialised, so we look at its
+        # spec instead
+        try:
+            orig = abspath(module.__spec__.origin)  # type: ignore
+        except (AttributeError, ValueError, TypeError):
+            orig = None
+
+    if orig is not None and isfile(orig):
+        if orig.endswith(".pyc"):
+            orig = orig[:-1]
+        return orig
+
+    return "<unknown origin>"
+
+
+def _resolve(path):
+    # type: (str) -> Optional[str]
+    """Resolve a (relative) path with respect to sys.path."""
+    for base in sys.path:
+        if isdir(base):
+            resolved_path = abspath(join(base, expanduser(path)))
+            if isfile(resolved_path):
+                return resolved_path
+    return None
+
+
+# Borrowed from the wrapt module
+# https://github.com/GrahamDumpleton/wrapt/blob/0baff1b6ac8d64ffb08ea2d1610c7ed90115b9d5/src/wrapt/importer.py
+class _ImportHookChainedLoader:
+    def __init__(self, loader, callback):
+        # type: (Loader, Callable[[ModuleType], None]) -> None
+        self.loader = loader
+        self.callback = callback
+
+    def load_module(self, fullname):
+        # type: (str) -> ModuleType
+        module = self.loader.load_module(fullname)
+        self.callback(module)
+
+        return module
+
+    def get_code(self, mod_name):
+        return self.loader.get_code(mod_name)
+
+
+class ModuleWatchdog(dict):
+    """Module watchdog.
+
+    Replace the standard ``sys.modules`` dictionary to detect when modules are
+    loaded/unloaded. This is also responsible for triggering any registered
+    import hooks.
+    """
+
+    _run_code = None
+
+    def __init__(self):
+        # type: () -> None
+        self._hook_map = defaultdict(list)  # type: DefaultDict[str, List[Tuple[HookType, Any]]]
+        self._origin_map = {origin(module): module for module in sys.modules.values()}
+        self._modules = sys.modules
+        self._finding = set()  # type: Set[str]
+
+    def __enter__(self):
+        # type: () -> ModuleWatchdog
+        # DEV: We need to patch __import__ because __setitem__ might be getting
+        # called too early and it might be empty or partially initialised.
+        self._patch_import()
+        sys.modules = self
+        log.debug("%s installed", type(self))
+        return self
+
+    def __exit__(self, *exc):
+        # type: (Tuple[Any]) -> None
+        self._unpatch_import()
+        sys.modules = self._modules
+        log.debug("%s uninstalled", type(self))
+
+    def __getitem__(self, item):
+        # type: (str) -> ModuleType
+        return self._modules.__getitem__(item)
+
+    def __setitem__(self, name, module):
+        # type: (str, ModuleType) -> None
+        self._modules.__setitem__(name, module)
+
+    def _patch_import(self):
+        # type: () -> None
+        if sys.meta_path and type(sys.meta_path[0]) is type(self):
+            # Already patched
+            return
+
+        sys.meta_path.insert(0, self)
```

Comment:
if you are just patching `sys.modules` do you need to have an import hook too?

Reply:
My initial attempt was based solely on patching, but it turned out not to be too reliable. Not all the imports were going through `__setitem__` and even patching `__import__` didn't work all the times. Besides, patching `__import__` is discouraged as per [PEP 302](https://www.python.org/dev/peps/pep-0302/). This hybrid approach is more for catching the `__delitem__` calls to ensure clean-up and avoid potential memory leaks.

Reply:
yeah, this is why in the last iteration I tried to do this I ended up patching a CPython internal function ("find_and_load" something), but that proofed to cause some issues.

an alternative approach I tried as well is to use the new `sys.audit` hooks. the problem is that the `import` event happens before the module is loaded (which makes sense since you can raise an error to prevent it from loading).

however, there are compile/code.__new__ events which you might be able to use instead.

of course this will only help on 3.8+, but maybe if it is more reliable then can relax the implementation for <3.8 to "best effort as CPython makes available to us"?

Reply:
Indeed having to support 2.7 and earlier Python 3 versions is a bit of a hindrance. The current approach seems to be quite solid for now, but as we drop support for legacy interpreters we should definitely consider better ways.